nifi-nar-bundles/nifi-smb-bundle/nifi-smb-processors/src/main/java/org/apache/nifi/processors/smb/FetchSmb.java

```diff
@@ -0,0 +1,201 @@
+            new Relationship.Builder()
+                    .name("success")
+                    .description("A flowfile will be routed here for each successfully fetched File.")
+                    .build();
+    public static final Relationship REL_FAILURE =
+            new Relationship.Builder().name("failure")
+                    .description(
+                            "A flowfile will be routed here for each File for which fetch was attempted but failed.")
+                    .build();
+    public static final Relationship REL_INPUT_FAILURE =
+            new Relationship.Builder().name("input_failure")
+                    .description("The incoming flowfile will be routed here if its content could not be processed.")
+                    .build();
+    public static final Set<Relationship> relationships = Collections.unmodifiableSet(new HashSet<>(asList(
+            REL_SUCCESS,
+            REL_FAILURE,
+            REL_INPUT_FAILURE
+    )));
+    static final PropertyDescriptor RECORD_READER = new PropertyDescriptor.Builder()
+            .name("record-reader")
+            .displayName("Record Reader")
+            .description(
+                    "Specifies the Controller Service to use for reading incoming NiFi Records. Each record should contain \"identifier\""
+                            + " attribute set to the path and name of the file to fetch."
+                            + " If not set, the Processor expects as attributes of a separate flowfile for each File to fetch.")
+            .identifiesControllerService(RecordReaderFactory.class)
+            .required(false)
+            .build();
+    private static final List<PropertyDescriptor> PROPERTIES = asList(
+            FILE_ID,
+            SMB_CLIENT_PROVIDER_SERVICE,
+            RECORD_READER
+    );
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        if (context.getProperty(RECORD_READER).isSet()) {
+            final RecordReaderFactory recordReaderFactory =
+                    context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+
+            try (InputStream inFlowFile = session.read(flowFile)) {
+                final Map<String, String> flowFileAttributes = flowFile.getAttributes();
+                final RecordReader
+                        reader =
+                        recordReaderFactory.createRecordReader(flowFileAttributes, inFlowFile, flowFile.getSize(),
+                                getLogger());
+
+                Record record;
+                while ((record = reader.nextRecord()) != null) {
+                    final String fileName = record.getAsString("identifier");
```

Comment:
I've quickly added the new property. Although the best would be to use EL in FILE_ID the same way I think. So that it would be possible to extend file names with a path prefix for example. I'm not sure if that's doable or should it be done in this PR.

Reply:
@exceptionfactory Good point! Record processing was introduced in `FetchGoogleDrive` for the first time and this use case was not handled. We just expect the value in a given field in the record and ignore the `File ID` property. The former can be documented but the latter is still weird. Adding a separate property for the record field/path would not help either because `File ID` would still be present but ignored.
Would it make sense to use the same property for both regular and record processing and expect literal/EL in the first case and a record path in the second case? (and provide documentation with examples)

Reply:
The main purpose of the inclusion of the record reader is to facilitate a more performing cooperation between the List- and the Fetch- processor.
There are multiple technical challenges that would need to be solved if we wanted to properly expose the record-based configuration to the user (among which the `File ID` is just one of many).

I think the best would be to leave it as it is with a small improvement. When an individual fetch - based on a single record - fails, it generates a flowfile with attributes what were record fields before. We could make it so that those flowfiles can be retried by the processor without stopping and reconfiguring.
We could see if the `File ID` attribute is present on the flowfile or not and based on that would we try to process the flowfile with the record reader or by just parsing the attributes.

This would also eliminate the problem where the `File ID` is not used when the record reader is set.